client/state/billing-transactions/actions.ts

```diff
@@ -28,28 +28,27 @@ export const requestBillingTransactions = ( transactionType?: BillingTransaction
 			const {
 				billing_history,
 				upcoming_charges,
+				billing_history_total,
 			}: {
-				billing_history: BillingTransaction[];
-				upcoming_charges: UpcomingCharge[];
+				billing_history?: BillingTransaction[];
+				upcoming_charges?: UpcomingCharge[];
+				billing_history_total?: number;
 			} = await wp.req.get( url, {
 				apiVersion: '1.3',
 			} );
-
-			const fullBillingHistory = billing_history;
-			if ( billing_history.length === limit ) {
-				// If we have exactly the limit number of transactions, it means there are more transactions to fetch.
-				let nextData = [];
+			const fullBillingHistory = billing_history ?? [];
+			if ( fullBillingHistory.length === limit && ( billing_history_total ?? 0 ) !== limit ) {
+				// If we have exactly the limit number of transactions (and this is not the full total), it means there are more transactions to fetch.
 				do {
 					// Fetch the next page of transactions.
-					const offset_url = url + '&offset=' + billing_history.length;
+					const offset_url = url + '&offset=' + fullBillingHistory.length;
 
 					const res = await wp.req.get( offset_url, {
```

Comment:
I suggest replacing `billing_history_total` here with the newly downloaded value, just in case the number of receipts has somehow decreased (I know it's very unlikely) since the first call, because if that happened then this would become an infinite loop.